meta/saiserializetest.c

```diff
@@ -1532,6 +1551,195 @@ void test_deserialize_pointer()
+    ASSERT_TRUE(res == -1, "expected true");
+}
+
+void test_serialize_attr_id()
+{
+    int res;
+    char buf[PRIMITIVE_BUFFER_SIZE];
+    sai_attr_id_t attr_id;
+
+    attr_id = SAI_SWITCH_ATTR_NUMBER_OF_ACTIVE_PORTS;
+
+    res = sai_serialize_attr_id(buf, NULL, attr_id);
+
+    ASSERT_STR_EQ(buf, "0", res);
+
+    /* test all ids */
+
+    size_t i = 0;
+
+    for (; i < sai_metadata_attr_sorted_by_id_name_count; ++i)
+    {
+        const sai_attr_metadata_t* amd = sai_metadata_attr_sorted_by_id_name[i];
+
+        res = sai_serialize_attr_id(buf, amd, amd->attrid);
+
+        ASSERT_STR_EQ(buf, amd->attridname, res);
+    }
+}
+
+void test_deserialize_attr_id()
+{
+    int res;
+    sai_attr_id_t attr_id;
+
+    res = sai_deserialize_attr_id("100", &attr_id);
+    ASSERT_TRUE(res == strlen("100"), "expected true");
+    ASSERT_TRUE(attr_id == 100, "expected true, value = %d", attr_id);
+
+    res = sai_deserialize_attr_id("-1", &attr_id);
+    ASSERT_TRUE(res == -1, "expected true");
+
+    /* test all enums */
+
+    size_t i = 0;
+
+    for (; i < sai_metadata_attr_sorted_by_id_name_count; ++i)
+    {
+        const sai_attr_metadata_t* amd = sai_metadata_attr_sorted_by_id_name[i];
+
+        res = sai_deserialize_attr_id(amd->attridname, &attr_id);
+
+        ASSERT_TRUE(res == (int)strlen(amd->attridname), "expected true");
+
+        ASSERT_TRUE(attr_id == amd->attrid, "expected true");
+    }
+}
+
+void test_serialize_attribute()
+{
+    int res;
+    char buf[PRIMITIVE_BUFFER_SIZE * 2];
+    sai_attribute_t attribute = {0};
+    const sai_attr_metadata_t* amd;
+
+    amd = sai_metadata_attr_by_object_type[SAI_OBJECT_TYPE_SWITCH][0];
+    attribute.id = SAI_SWITCH_ATTR_NUMBER_OF_ACTIVE_PORTS;
+    attribute.value.u32 = 3;
+
+    res = sai_serialize_attribute(buf, amd, &attribute);
+
+    ASSERT_STR_EQ(buf, "{\"id\":\"SAI_SWITCH_ATTR_NUMBER_OF_ACTIVE_PORTS\",\"value\":{\"u32\":3}}", res);
+
+    attribute.id = 99999;
+    memset(&attribute.value, 0x55, sizeof(attribute.value));
+
+    res = sai_serialize_attribute(buf, NULL, &attribute);
+
+    ASSERT_STR_EQ(buf, "{\"id\":\"99999\",\"value\":{85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85}}", res);
+}
+
+void test_deserialize_attribute()
+{
+    int res;
+    const char *buf;
+    sai_attribute_t attribute = {0};
+    sai_attribute_value_t value;
+
+    buf = "{\"id\":\"SAI_SWITCH_ATTR_NUMBER_OF_ACTIVE_PORTS\",\"value\":{\"u32\":3}}";
+    res = sai_deserialize_attribute(buf, &attribute);
+    ASSERT_TRUE(res == (int)strlen(buf), "expected true");
+    ASSERT_TRUE(attribute.id == SAI_SWITCH_ATTR_NUMBER_OF_ACTIVE_PORTS, "expected true");
+    ASSERT_TRUE(attribute.value.u32 == 3, "expected true");
+
+    buf = "{\"id\":\"99999\",\"value\":{85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85,85}}";
```

Comment:
serialize/deserialize attribute IDs which are not valid attributes that are not defined in sai headers make no sense

Reply:
Invalid attribute IDs maybe passed to SAI under some error condition. SAI code of course will do sanity check and return failure. In this case, it will be nice if SAI code can serialize and record the invalid IDs.

Reply:
what conditions? if attribute id is invalid it will be logged by warning message to syslog, thats where you get actual id which is invalid, no need to actually serialize invalid id

Reply:
Two possible conditions: 1. Code error in NOS 2. Test cases which are checking if SAI code can deal with the invalid id in right way.
Yes, the invalid id can also be logged by warning messages. However, there are still benefits if the invalid id can be serialized. Firstly, this makes sure the serializing code does not crush under this condition, the NOS may have a chance to know the id is invalid by checking SAI return values. Secondly, this makes the serialized data complete (SAI can serialize and record all the API calling data), which is good for debugging SAI related issues.